contrib/babelfishpg_tsql/src/tsql_analyze.c

```diff
@@ -313,34 +317,186 @@ pre_transform_setop_tree(SelectStmt *stmt, SelectStmt *leftmostSelect)
 		return;
 
 	leftmostSelect->sortClause = stmt->sortClause;
-	stmt->sortClause = NIL;
 }
 
 /* 
- * After the sort clause has been analyzed in the leftmost select, update the 
- * top-level TLEs to contain a ref to the sortlist and restore the sortClause.
+ * fix_setop_typmods will backtrack through the SetOperationTree to
+ * fix the types of char, nchar, varchar, and nvarchar
+ *
+ * We do this after the tree has already been transformed so we may 
+ * compare all expressions in the same column at once. PG only compares
+ * two at a time, leading to issues for example if both expressions are NULL.
  * 
- * We can trust that the ordering of TLEs matches as PG uses the leftmost 
- * target list as a basis for builiding the top-level target list, it just 
- * doesn't fill in the ressortgroupref field. 
+ * Then, we update the original expressions as well as the top-level target list's
+ * expressions with the correct type and typmod. 
+ */
+static void
+fix_setop_typmods(ParseState *pstate, Query *qry)
+{
+	List 		*setOpTreeStack = list_make1(qry->setOperations);
+	List		*setOpNodes = NIL;
+	List 		*collist_list = NIL;
+	List		*topColTypes = NIL;
+	List		*topColTypmods = NIL;
+	List		*topColCollations = NIL;
+	ListCell	*collistl, *setopsl, *toptlistl;
+	Oid 		common_type, common_collation;
+	int32 		common_typmod;
+
+	/* Iterate through the SetOpTree. For each column, save each expression
+	 * in that column to a list. That is, for select a, b, c union select x, y, x,
+	 * give [a, x], [b, y], [c, z] 	*/
+	while (setOpTreeStack)
+	{
+		Node *setOp = llast(setOpTreeStack);
+		setOpTreeStack = list_delete_last(setOpTreeStack);
+
+		if (IsA(setOp, SetOperationStmt))
+		{
+			SetOperationStmt *op = (SetOperationStmt *) setOp;
+			setOpNodes = lappend(setOpNodes, op);
+			setOpTreeStack = lappend(setOpTreeStack, op->rarg);
+			setOpTreeStack = lappend(setOpTreeStack, op->larg);
+		} else if (IsA(setOp, RangeTblRef))
+		{
+			RangeTblEntry *rte = rt_fetch(((RangeTblRef*)setOp)->rtindex, pstate->p_rtable);
+			List *targetList = rte->subquery->targetList;
```

Comment:
will there always be subtree? does it make sense to check it before accessing it?